src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -226,5 +226,90 @@ public static object InvokeMethod( MethodInfo method, object fixture, params obj
+        /// <param name="name">See <see cref="TypeInfo.GetProperty(string, BindingFlags)"/>.</param>
+        /// <param name="bindingFlags">See <see cref="TypeInfo.GetProperty(string, BindingFlags)"/>.</param>
+#else
+        /// <summary>
+        /// <para>
+        /// Selects the ultimate shadowing property just like <see langword="dynamic"/> would,
+        /// rather than throwing <see cref="AmbiguousMatchException"/>
+        /// for properties that shadow properties of a different property type
+        /// which is what <see cref="Type.GetProperty(string, BindingFlags)"/> does.
+        /// </para>
+        /// <para>
+        /// If you request both public and nonpublic properties, every public property is preferred
+        /// over every nonpublic property. It would violate the principle of least surprise for a
+        /// derived class�s implementation detail to be chosen over the public API for a type.
+        /// </para>
+        /// </summary>
+        /// <param name="type">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
+        /// <param name="name">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
+        /// <param name="bindingFlags">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
+#endif
+
+        public static PropertyInfo GetUltimateShadowingProperty(Type type, string name, BindingFlags bindingFlags)
+        {
+            if (type == null) throw new ArgumentNullException(nameof(type));
+            if (name == null) throw new ArgumentNullException(nameof(name));
+
+            if ((bindingFlags & BindingFlags.DeclaredOnly) != 0)
+            {
+                // If you're asking us to search a hierarchy but only want properties declared in the given type,
+                // you're in the wrong place but okay:
+                return type.GetProperty(name, bindingFlags);
+            }
+
+            if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == (BindingFlags.Public | BindingFlags.NonPublic))
+            {
+                // If we're searching for both public and nonpublic properties, search for only public first
+                // because chances are if there is a public property, it would be very surprising to detect the private shadowing property.
+
+                for (var publicSearchType = type; publicSearchType != null; publicSearchType = publicSearchType.GetTypeInfo().BaseType)
+                {
+                    var property = publicSearchType.GetProperty(name, (bindingFlags | BindingFlags.DeclaredOnly) & ~BindingFlags.NonPublic);
+                    if (property != null) return property;
+                }
+
+                // There is no public property, so may as well not ask to include them during the second search.
+                bindingFlags &= ~BindingFlags.Public;
+            }
+
+            for (; type != null; type = type.GetTypeInfo().BaseType)
```

Comment:
More out of interest - what's going on here? 🙂 i.e. why half a for loop, and not something more explicit. I've no doubt you have a good reason!

Reply:
I guess I had better write it like the other loop then. It's doing the exact same iteration, only it's using `type` as the iteration variable rather than introducing a new `searchType` variable.
I wrote this loop first but I should come back and make it consistent with the one in the `if` block.

Reply:
Ah - it's working up the tree? Clever!

Reply:
Yeah, it's the linked-list style of `for` loop rather than the array style.